compute/src/test/java/org/jclouds/compute/domain/internal/ArbitraryCpuRamTemplateBuilderImplTest.java

```diff
@@ -0,0 +1,142 @@
+      Provider<TemplateOptions> optionsProvider = createMock(Provider.class);
+      Provider<TemplateBuilder> templateBuilderProvider = createMock(Provider.class);
+      TemplateBuilder defaultTemplate = createMock(TemplateBuilder.class);
+      TemplateOptions defaultOptions = createMock(TemplateOptions.class);
+      GetImageStrategy getImageStrategy = createMock(GetImageStrategy.class);
+
+      expect(optionsProvider.get()).andReturn(new TemplateOptions());
+      expect(getImageStrategy.getImage(anyObject(String.class))).andReturn(null);
+      replay(defaultTemplate, optionsProvider, templateBuilderProvider, getImageStrategy);
+
+      TemplateBuilderImpl templateBuilder = new ArbitraryCpuRamTemplateBuilderImpl(locations, new ImageCacheSupplier(images, 60,
+            Atomics.<AuthorizationException>newReference(), Providers.of(getImageStrategy)), hardwares,
+            Suppliers.ofInstance(region),
+            optionsProvider, templateBuilderProvider);
+
+      Hardware hardware = templateBuilder.hardwareId("automatic:cores=2;ram=256").build().getHardware();
+      assertThat(hardware.getRam()).isEqualTo(256);
+      assertThat(hardware.getProcessors()).extracting("cores").containsExactly(2.0);
+   }
+
+   @Test
+   public void testAutoGeneratedHardwareWithMinCoresAndMinRamTest(){
+      Supplier<Set<? extends Location>> locations = Suppliers.<Set<? extends Location>> ofInstance(ImmutableSet.of(region));
+      Supplier<Set<? extends Image>> images = Suppliers.<Set<? extends Image>> ofInstance(ImmutableSet.of(image));
+      Supplier<Set<? extends Hardware>> hardwares = Suppliers.<Set<? extends Hardware>> ofInstance(ImmutableSet
+            .<Hardware> of(createMock(Hardware.class)));
+      Provider<TemplateOptions> optionsProvider = createMock(Provider.class);
+      Provider<TemplateBuilder> templateBuilderProvider = createMock(Provider.class);
+      TemplateBuilder defaultTemplate = createMock(TemplateBuilder.class);
+      TemplateOptions defaultOptions = createMock(TemplateOptions.class);
+      GetImageStrategy getImageStrategy = createMock(GetImageStrategy.class);
+
+      expect(optionsProvider.get()).andReturn(new TemplateOptions());
+      expect(getImageStrategy.getImage(anyObject(String.class))).andReturn(null);
+      replay(defaultTemplate, optionsProvider, templateBuilderProvider, getImageStrategy);
+      TemplateBuilderImpl templateBuilder = new ArbitraryCpuRamTemplateBuilderImpl(locations, new ImageCacheSupplier(images, 60,
+            Atomics.<AuthorizationException>newReference(), Providers.of(getImageStrategy)), hardwares,
+            Suppliers.ofInstance(region), optionsProvider, templateBuilderProvider);
+      templateBuilder.minRam(1024);
+      templateBuilder.minCores(2);
+      Hardware hardware = templateBuilder.build().getHardware();
+      assertThat(hardware.getRam()).isEqualTo(1024);
+      assertThat(hardware.getProcessors()).extracting("cores").containsExactly(2.0);
+   }
```

Comment:
Add a test to verify what happens when you only set the minRam or the minCores, as it is a path that your class is explicitly handling.
Also add a couple more tests where there is actually a hardware profile present:

* One test where the existing hardware profile does not match, and an automatic one is used.
* One test where the existing hardware profile matches the given cpu and ram, to verify that the existing hardware profile is used.